codegen/src/main/mod.rs

```diff
@@ -0,0 +1,390 @@
+            fn_inputs: vec![
+                Input {
+                    ident: parse_quote!(pool),
+                    builder: parse_quote!(shared::Postgres),
+                },
+                Input {
+                    ident: parse_quote!(redis),
+                    builder: parse_quote!(shared::Redis),
+                },
+            ],
+        };
+
+        let actual = quote!(#input);
+        let expected = quote! {
+            async fn __shuttle_wrapper(
+                factory: &mut dyn shuttle_service::Factory,
+                runtime: &shuttle_service::Runtime,
+                logger: Box<dyn shuttle_service::log::Log>,
+            ) -> Result<Box<dyn shuttle_service::Service>, shuttle_service::Error> {
+                use shuttle_service::ResourceBuilder;
+
+                runtime.spawn_blocking(move || {
+                    shuttle_service::log::set_boxed_logger(logger)
+                        .map(|()| shuttle_service::log::set_max_level(shuttle_service::log::LevelFilter::Info))
+                        .expect("logger set should succeed");
+                })
+                .await
+                .map_err(|e| {
+                    if e.is_panic() {
+                        let mes = e
+                            .into_panic()
+                            .downcast_ref::<&str>()
+                            .map(|x| x.to_string())
+                            .unwrap_or_else(|| "<no panic message>".to_string());
+
+                        shuttle_service::Error::BuildPanic(mes)
+                    } else {
+                        shuttle_service::Error::Custom(shuttle_service::error::CustomError::new(e))
+                    }
+                })?;
+
+                let pool = shuttle_service::shared::Postgres::new().build(factory, runtime).await?;
+                let redis = shuttle_service::shared::Redis::new().build(factory, runtime).await?;
+
+                runtime.spawn(async {
+                    complex(pool, redis)
+                        .await
+                        .map(|ok| Box::new(ok) as Box<dyn shuttle_service::Service>)
+                })
+                .await
+                .map_err(|e| {
+                    if e.is_panic() {
+                        let mes = e
+                            .into_panic()
+                            .downcast_ref::<&str>()
+                            .map(|x| x.to_string())
+                            .unwrap_or_else(|| "<no panic message>".to_string());
+
+                        shuttle_service::Error::BuildPanic(mes)
+                    } else {
+                        shuttle_service::Error::Custom(shuttle_service::error::CustomError::new(e))
+                    }
+                })?
+            }
+        };
+
+        assert_eq!(actual.to_string(), expected.to_string());
+    }
+
+    #[test]
+    fn ui() {
+        let t = trybuild::TestCases::new();
+        t.compile_fail("tests/ui/*.rs");
```

Comment:
Should these maybe be moved too? Maybe something like `tests/ui/main/*.rs` :man_shrugging:

Reply:
Sure! So you're thinking that they would be run as integration tests rather than unit tests?

Reply:
For these, they can be integration or unit tests really - can't quite remember why I made them unit tests :smile: 

I was thinking more of moving the `ui` files since another set will eventually be added to test the codegen for the lambda macro

Reply:
Oh, I see what you mean. Sure! That makes sense